### ISSUE TITLE
Client -> Server error Server returned nothing (no headers, no data)

### ISSUE THREAD

**Issue by MHamburg** [freelan-developers/freelan#84]
Currently I am using the built-in webserver for a multi node registration setup. While the following setup works on the localhost (both client and server). A remote host is unable to connect, manually I can connect (using fiddler).

Environment:
WIndows 2012 R2

[Node 1] [Config]
[server]
enabled=yes
listen_on=0.0.0.0:443
protocol=https
server_certificate_file=crt\server.crt
server_private_key_file=key\server.key
certification_authority_certificate_file=crt\ca.crt
certification_authority_private_key_file=key\ca.key
authentication_script="C:\Program Files\FreeLAN\config\auth\FreeLanAuth.exe"

#connect to self
[client]
enabled=no
server_endpoint=127.0.0.1
disable_peer_verification=yes
disable_host_verification=yes
username=Test
password=Test123

[Node 2] [Config]
[client]
enabled=yes
server_endpoint=a.com:443
disable_peer_verification=yes
disable_host_verification=yes
username=Test
password=Test123

Error:
[Client]
2015-07-14T20:14:56.090563 [INFORMATION] Requesting a client certificate from th
e web server...
2015-07-14T20:14:57.137442 [ERROR] Error while sending HTTP(S) request to https:
//a.com:443/request_certificate/: Server returned nothing (no header
s, no data) (curl::error:52)
2015-07-14T20:14:57.153063 [ERROR] The certificate request to the web server fai
led: Server returned nothing (no headers, no data) (curl::error:52). Retrying in
 00:01:30...
2015-07-14T20:15:08.012443 [INFORMATION] Requesting the CA certificate from the
web server...
2015-07-14T20:15:08.528068 [ERROR] Error while sending HTTP(S) request to https:
//a.com:443/request_ca_certificate/: Server returned nothing (no hea
ders, no data) (curl::error:52)
2015-07-14T20:15:08.543694 [ERROR] The CA certificate request to the web server
failed: Server returned nothing (no headers, no data) (curl::error:52). Retrying
 in 00:01:30...

[Server]
 2015-07-15T03:17:38.563635 [INFORMATION] Web server - Received GET request from
213.125.129.122:50790 for /request_ca_certificate/ (0 byte(s) content).
2015-07-15T03:17:38.563635 [DEBUG] --- Headers follow ---
2015-07-15T03:17:38.579264 [DEBUG] authorization: Basic VGVzdDpUZXN0MTIz
2015-07-15T03:17:38.579264 [DEBUG] host: a.com
2015-07-15T03:17:38.579264 [DEBUG] accept: */*
2015-07-15T03:17:38.579264 [DEBUG] --- End of headers ---
2015-07-15T03:17:38.579264 [DEBUG] Calling script C:\Program Files\FreeLAN\confi
g\auth\FreeLanAuth.exe...
2015-07-15T03:17:38.626140 [DEBUG] Script C:\Program Files\FreeLAN\config\auth\F
reeLanAuth.exe returned 0.
2015-07-15T03:17:38.641766 [DEBUG] Authentication succeeded for user "Test".
2015-07-15T03:17:38.641766 [DEBUG] Test (123.125.111.000:50790) requested the CA
 certificate.
 
[Fiddler]
 Exact same request was made, from the exact same client machine, to request_ca_certificate/:
 HTTP/1.1 200 OK
set-cookie: session_id=KNmiDJe1I20HqEw6a64pZJWkPqoutdsrEU5FFwCiHEY=; Expires=Wed, 15 Jul 2015 03:08:44 GMT; HttpOnly
content-type: application/x-x509-ca-cert
Transfer-Encoding: chunked

result in Overall Elapsed:	0:00:00.218

**Comment by ereOn** [comment]
@MHamburg Sorry, I was offline for a while during the summer and I just realized I forgot this thread.

It is unclear to me what happens in the last case: does Fiddler show that the server indeed replies with some content when you request the `request_ca_certificate/` URL ?

**Comment by MHamburg** [comment]
@ereOn 
 Please find my respons below:

does Fiddler show that the server indeed replies with some content when you request the request_ca_certificate/ URL ? 

- Yes the certificate is returned in binary form

Did you run fiddler from a remote host as well ? 

- Yes you can do the same if you wish (it is a sandbox setup and will be removed later on):

Request GET: https://191.238.149.167/request_ca_certificate/
Auth-Header: authorization: Basic VGVzdDpUZXN0MTIz

**Comment by ereOn** [comment]
@MHamburg That is super weird.

Would you happen to be behind a proxy ?

Could you run wireshark on the client to look at what the client/server sends/replies when that happens ?

**Comment by leggewie** [comment]
@MHamburg Thank you for reporting this issue

unblocking for 2.1 until the issue has been successfully reproduced by a developer.

@ereOn What is the future for the web server support?

**Comment by ereOn** [comment]
@leggewie Last time I checked, implementing a reasonably compliant web server in C++ was a lot of unpleasant work.

I have been toying around a few techs these days and I might come up with a rather "original" idea. I need more time to assess it's feasability though.

**Comment by leggewie** [comment]
I'm OK with dropping that feature in favor of making the core function awesome and especially making the general setup drop-dead easy.  There's work underway and #144 was a good step in that direction, for example.  I hope we can release 2.1 soon.

My vote goes to closing this ticket as out of scope and drop the functionality.

**Comment by s-vincent** [comment]
If we take into account that mongoose (library used for the web server feature) is disabled by default and violates GPLv3 (so we cannot ship binary), I am also in favor of dropping it.

But I think for 2.1 we need either a replacement feature for that or some kind of super-node feature (based on ICE protocol for example) or the mysterious "original" idea from ereOn :).

**Comment by leggewie** [comment]
What is the web server doing that can't be (easily) done otherwise ATM?

**Comment by leggewie** [comment]
I've added a pull request to add an FAQ to the website for a drop-dead easy configuration of freelan.  The only requirement still present would be for at least one of the VPN nodes to have one UDP port exposed to the web.  That's really not much different from the requirement for the web server that also needs one port (TCP) exposed.

https://github.com/freelan-developers/website/pull/5

**Comment by s-vincent** [comment]
@ereOn 
> I have been toying around a few techs these days and I might come up with a rather "original" idea. I need more time to assess it's feasability though.


Any news on that  "original idea" ?

**Comment by s-vincent** [comment]
@ereOn, can you please state about this one ?

> I have been toying around a few techs these days and I might come up with a rather "original" idea. I > need more time to assess it's feasability though.

In parallel, I heard about Boost.Beast which can do "low-level" HTTP in Boost.ASIO. Maybe it is a good candidate to replace mongoose. The only drawback I see is that it requires boost 1.66 (so Debian Stretch and CentOS are out of the game :/).

What do you think?

**Comment by s-vincent** [comment]
@MHamburg Now that use vcpkg for Windows dependencies, the web client library we use (curl/openssl) stuff is quite up-to-date.

Can you please give a try on latest git revision? You will need to patch the libs/libfreelan/libfreelan.vcxproj to add USE_MONGOOSE define in <PreprocessorDefinitions> section to enable server feature.

**Comment by ereOn** [comment]
@s-vincent Sorry for the delay, I'm a bit behind my notifications these days :P (catching up with November at the moment).

My "original" idea was to toy around a Go implementation. Go makes it incredibly easy to host prodution-grade HTTP(S) servers.

As I was playing with that, I realized reimplementing freelan itself in Go would have a lot of benefits, namely:

- Greater portability.
- Lower developement entry-skills-level.
- Less requirements for external libraries (everything network & crypto comes with the standard library in Go...).

Obviously, this is major and I'm not even sure I want to go that way at all. As I said, I'm toying with the idea and almost have a fully compliant implementation. That sadly, doesn't solve the problem for the current existing and authoritative C++ implementation.

I wonder how widespread the usage for this HTTP server is, out there in the wild. Is it even worth keeping it as it is compared the maintainability burden?
